options/logger.go

```diff
@@ -78,194 +52,264 @@ func (f LogFormat) MakeFormatter() (send.MessageFormatter, error) {
+type RawLoggerConfigFormat string
+
+const (
+	RawLoggerConfigFormatBSON    RawLoggerConfigFormat = "BSON"
+	RawLoggerConfigFormatJSON    RawLoggerConfigFormat = "JSON"
+	RawLoggerConfigFormatInvalid RawLoggerConfigFormat = "invalid"
+)
+
+// Validate ensures that RawLoggerConfigFormat is valid.
+func (f RawLoggerConfigFormat) Validate() error {
+	switch f {
+	case RawLoggerConfigFormatBSON, RawLoggerConfigFormatJSON:
+		return nil
+	case RawLoggerConfigFormatInvalid:
+		return errors.New("invalid log format")
+	default:
+		return errors.Errorf("unknown raw logger config format '%s'", f)
+	}
 }
 
-// Validate ensures that BufferOptions is valid.
-func (opts Buffer) Validate() error {
-	if opts.Buffered && opts.Duration < 0 || opts.MaxSize < 0 {
-		return errors.New("cannot have negative buffer duration or size")
+func (f RawLoggerConfigFormat) unmarshal(data []byte, out LoggerProducer) error {
+	switch f {
+	case RawLoggerConfigFormatBSON:
+		if err := bson.Unmarshal(data, out); err != nil {
+			return errors.Wrapf(err, "could not render '%s' input into '%s'", data, out)
+
+		}
+
+		return nil
+	case RawLoggerConfigFormatJSON:
+		if err := json.Unmarshal(data, out); err != nil {
+			return errors.Wrapf(err, "could not render '%s' input into '%s'", data, out)
+
+		}
+
+		return nil
+	default:
+		return errors.Errorf("unsupported format '%s'", f)
 	}
-	return nil
 }
 
-// Log contains options related to the logging done by Jasper.
-//
-// By default, logger reads from both standard output and standard error.
-type Log struct {
-	BufferOptions        Buffer                    `json:"buffer_options,omitempty" bson:"buffer_options"`
-	BuildloggerOptions   send.BuildloggerConfig    `json:"buildlogger_options,omitempty" bson:"buildlogger_options"`
-	BuildloggerV3Options timber.LoggerOptions      `json:"buildlogger_v3_options" bson:"buildlogger_v_3_options"`
-	DefaultPrefix        string                    `json:"default_prefix,omitempty" bson:"default_prefix"`
-	FileName             string                    `json:"file_name,omitempty" bson:"file_name"`
-	Format               LogFormat                 `json:"format" bson:"format"`
-	InMemoryCap          int                       `json:"in_memory_cap,omitempty" bson:"in_memory_cap"`
-	Level                send.LevelInfo            `json:"level,omitempty" bson:"level"`
-	SplunkOptions        send.SplunkConnectionInfo `json:"splunk_options,omitempty" bson:"splunk_options"`
-	SumoEndpoint         string                    `json:"sumo_endpoint,omitempty" bson:"sumo_endpoint"`
+// LoggerConfig represents the necessary information to construct a new grip
+// send.Sender. LoggerConfig implements the bson.MarshalBSON and
+// json.MarshalJSON interfaces.
+type LoggerConfig struct {
+	Type     string                `json:"type" bson:"type"`
+	Format   RawLoggerConfigFormat `json:"format" bson:"format"`
+	Config   []byte                `json:"config" bson:"config"`
```

Comment:
I think that this needs to be a type alias with marshal functions so that when we marshal the logger config, this doesn't end up coming out as numbers. Putting the on the LoggerConfig object means that you have to duplicate the structure and tags in an autonomous struct, and if you put it here, and produce the byte slice in Set() then it's possible to modify LoggerConfig in the future (if we want/need to) without needing to modify the marshaling functions.